virttest/env_process.py

```diff
@@ -319,32 +318,60 @@ def postprocess_image(test, params, image_name, vm_process_status=None):
             elif clone_master == "yes":
                 if image_name in params.get("master_images_clone").split():
                     image.check_image(params, base_dir, force_share=True)
-            # Allow test to overwrite any pre-testing  automatic backup
-            # with a new backup. i.e. assume pre-existing image/backup
-            # would not be usable after this test succeeds. The best
-            # example for this is when 'unattended_install' is run.
-            if params.get("backup_image", "no") == "yes":
-                image.backup_image(params, base_dir, "backup", True)
-                restored = True
-            elif params.get("restore_image", "no") == "yes":
-                image.backup_image(params, base_dir, "restore", True)
-                restored = True
         except Exception as e:
-            if params.get("restore_image_on_check_error", "no") == "yes":
-                image.backup_image(params, base_dir, "restore", True)
-            if params.get("remove_image_on_check_error", "no") == "yes":
-                cl_images = params.get("master_images_clone", "")
-                if image_name in cl_images.split():
-                    image.remove()
             if (params.get("skip_cluster_leak_warn") == "yes" and
                     "Leaked clusters" in e.message):
                 logging.warn(e.message)
             else:
+                # FIXME: remove it from params, maybe as an img object attr
+                params["img_check_failed"] = "yes"
                 raise e
+
+
+def postprocess_image(test, params, image_name, vm_process_status=None):
+    """
+    Postprocess a single QEMU image according to the instructions in params.
+
+    The main operation is to remove images if instructions are gaven.
+
+    :param test: An Autotest test object.
+    :param params: A dict containing image postprocessing parameters.
+    :param vm_process_status: (optional) vm process status like running, dead
+                              or None for no vm exist.
+    """
+    restored, removed = (False, False)
+    clone_master = params.get("clone_master", None)
+    base_dir = data_dir.get_data_dir()
+    image = qemu_storage.QemuImg(params, base_dir, image_name)
+    if params.get("img_check_failed") == "yes":
+        if params.get("restore_image_on_check_error", "no") == "yes":
+            image.backup_image(params, base_dir, "restore", True)
+            restored = True
+    else:
+        # Allow test to overwrite any pre-testing  automatic backup
+        # with a new backup. i.e. assume pre-existing image/backup
+        # would not be usable after this test succeeds. The best
+        # example for this is when 'unattended_install' is run.
+        if params.get("backup_image", "no") == "yes":
+            image.backup_image(params, base_dir, "backup", True)
```

Comment:
Should skip the restore operation when vm is running.

Reply:
all updated, thanks